File: Solutions/DailyCodingProblems/DCP-1289/Solution.py
from typing import Dict, List

def pagerank(
    graph: Dict[str, List[str]], 
    d: float = 0.85, 
    tol: float = 1e-6, 
    max_iter: int = 100
) -> Dict[str, float]:
    """
    Calculates the PageRank score for each website in a directed graph.
    
    :param graph: A dictionary representing the directed graph where keys are source
                  nodes and values are lists of target nodes.
    :param d: The damping factor (probability of continuing to click links).
    :param tol: The convergence tolerance threshold.
    :param max_iter: The maximum number of iterations.
    :return: A dictionary mapping each node to its PageRank score.
    """
    # Collect all unique nodes in the graph
    nodes = set(graph.keys())
    for targets in graph.values():
        nodes.update(targets)
        
    N = len(nodes)
    if N == 0:
        return {}
        
    # Initialize all ranks to 1/N
    ranks = {node: 1.0 / N for node in nodes}
    
    # Calculate outdegree for all nodes
    outdegree = {node: len(graph.get(node, [])) for node in nodes}
    
    # Calculate incoming links for all nodes
    incoming = {node: set() for node in nodes}
    for src, targets in graph.items():
        for dst in targets:
            incoming[dst].add(src)
            
    # Identify dangling nodes (outdegree = 0)
    dangling_nodes = [node for node in nodes if outdegree[node] == 0]
    
    for _ in range(max_iter):
        new_ranks = {}
        # Sum of ranks of all dangling nodes
        dangling_sum = sum(ranks[node] for node in dangling_nodes)
        
        for node in nodes:
            # Base (1 - d) / N + d * (rank of dangling nodes redistributed equally)
            rank_sum = dangling_sum / N
            
            # Add contributions from incoming pages
            for src in incoming[node]:
                rank_sum += ranks[src] / outdegree[src]
                
            new_ranks[node] = (1.0 - d) / N + d * rank_sum
            
        # Check convergence
        diff = max(abs(new_ranks[node] - ranks[node]) for node in nodes)
        ranks = new_ranks
        if diff < tol:
            break
            
    return ranks

if __name__ == "__main__":
    # Test Case 1: Simple cycle A <-> B
    print("Test 1: Simple Cycle (A <-> B)")
    g1 = {
        "A": ["B"],
        "B": ["A"]
    }
    r1 = pagerank(g1)
    for node, rank in sorted(r1.items()):
        print(f"  {node}: {rank:.4f}")
        
    # Test Case 2: Chain with a dangling node A -> B -> C
    print("\nTest 2: Chain (A -> B -> C)")
    g2 = {
        "A": ["B"],
        "B": ["C"]
    }
    r2 = pagerank(g2)
    for node, rank in sorted(r2.items()):
        print(f"  {node}: {rank:.4f}")

    # Test Case 3: Star Graph (A links to B, C, D)
    print("\nTest 3: Star Graph (A -> B, C, D)")
    g3 = {
        "A": ["B", "C", "D"]
    }
    r3 = pagerank(g3)
    for node, rank in sorted(r3.items()):
        print(f"  {node}: {rank:.4f}")

    # Test Case 4: Disconnected components (A -> B) and (C -> D)
    print("\nTest 4: Disconnected Components (A -> B, C -> D)")
    g4 = {
        "A": ["B"],
        "C": ["D"]
    }
    r4 = pagerank(g4)
    for node, rank in sorted(r4.items()):
        print(f"  {node}: {rank:.4f}")

File: Solutions/DailyCodingProblems/DCP-1289/Solution.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Solution
{
    public static Dictionary<string, double> PageRank(
        Dictionary<string, List<string>> graph, 
        double d = 0.85, 
        double tol = 1e-6, 
        int maxIter = 100)
    {
        // Gather all unique nodes in the graph
        var nodes = new HashSet<string>(graph.Keys);
        foreach (var targets in graph.Values)
        {
            foreach (var node in targets)
            {
                nodes.Add(node);
            }
        }

        int N = nodes.Count;
        if (N == 0) return new Dictionary<string, double>();

        // Initialize ranks to 1/N
        var ranks = nodes.ToDictionary(node => node, node => 1.0 / N);
        
        // Compute outdegrees
        var outdegree = nodes.ToDictionary(
            node => node, 
            node => graph.ContainsKey(node) ? graph[node].Count : 0
        );
        
        // Map incoming links
        var incoming = nodes.ToDictionary(node => node, node => new HashSet<string>());
        foreach (var kvp in graph)
        {
            string src = kvp.Key;
            foreach (var dst in kvp.Value)
            {
                incoming[dst].Add(src);
            }
        }

        // Identify dangling nodes
        var danglingNodes = nodes.Where(node => outdegree[node] == 0).ToList();

        for (int iter = 0; iter < maxIter; iter++)
        {
            var newRanks = new Dictionary<string, double>();
            double danglingSum = danglingNodes.Sum(node => ranks[node]);

            foreach (var node in nodes)
            {
                double rankSum = danglingSum / N;

                foreach (var src in incoming[node])
                {
                    rankSum += ranks[src] / outdegree[src];
                }

                newRanks[node] = (1.0 - d) / N + d * rankSum;
            }

            // Check convergence
            double maxDiff = nodes.Max(node => Math.Abs(newRanks[node] - ranks[node]));
            ranks = newRanks;

            if (maxDiff < tol)
            {
                break;
            }
        }

        return ranks;
    }

    static void Main()
    {
        // Test Case 1: Simple cycle A <-> B
        Console.WriteLine("Test 1: Simple Cycle (A <-> B)");
        var g1 = new Dictionary<string, List<string>>
        {
            { "A", new List<string> { "B" } },
            { "B", new List<string> { "A" } }
        };
        var r1 = PageRank(g1);
        foreach (var kvp in r1.OrderBy(x => x.Key))
        {
            Console.WriteLine($"  {kvp.Key}: {kvp.Value:F4}");
        }

        // Test Case 2: Chain with a dangling node A -> B -> C
        Console.WriteLine("\nTest 2: Chain (A -> B -> C)");
        var g2 = new Dictionary<string, List<string>>
        {
            { "A", new List<string> { "B" } },
            { "B", new List<string> { "C" } }
        };
        var r2 = PageRank(g2);
        foreach (var kvp in r2.OrderBy(x => x.Key))
        {
            Console.WriteLine($"  {kvp.Key}: {kvp.Value:F4}");
        }

        // Test Case 3: Star Graph (A links to B, C, D)
        Console.WriteLine("\nTest 3: Star Graph (A -> B, C, D)");
        var g3 = new Dictionary<string, List<string>>
        {
            { "A", new List<string> { "B", "C", "D" } }
        };
        var r3 = PageRank(g3);
        foreach (var kvp in r3.OrderBy(x => x.Key))
        {
            Console.WriteLine($"  {kvp.Key}: {kvp.Value:F4}");
        }

        // Test Case 4: Disconnected components (A -> B) and (C -> D)
        Console.WriteLine("\nTest 4: Disconnected Components (A -> B, C -> D)");
        var g4 = new Dictionary<string, List<string>>
        {
            { "A", new List<string> { "B" } },
            { "C", new List<string> { "D" } }
        };
        var r4 = PageRank(g4);
        foreach (var kvp in r4.OrderBy(x => x.Key))
        {
            Console.WriteLine($"  {kvp.Key}: {kvp.Value:F4}");
        }
    }
}
